jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java

```diff
@@ -0,0 +1,490 @@
+            }
+        }
+    }
+
+    private boolean doProduce(boolean nonBlocking)
+    {
+        Runnable task = produceTask();
+
+        if (task == null)
+        {
+            try (AutoLock l = _lock.lock())
+            {
+                // Could another task just have been queued with a produce call?
+                switch (_state)
+                {
+                    case PRODUCING:
+                        _state = State.IDLE;
+                        return false;
+
+                    case REPRODUCING:
+                        _state = State.PRODUCING;
+                        return true;
+
+                    default:
+                        throw new IllegalStateException(toStringLocked());
+                }
+            }
+        }
+
+        Mode mode;
+        // Switch on the invocation type of the produced task
+        Invocable.InvocationType taskType = Invocable.getInvocationType(task);
+        switch (taskType)
+        {
+            case NON_BLOCKING:
+                // The produced task will never block, so we directly consume the task with PC
+                // and then resume production.
+                mode = Mode.PRODUCE_CONSUME;
+                break;
+
+            case BLOCKING:
+            case EITHER:
+            default:
+                // if this thread may block then we need to take the lock so we can
+                // atomically check/mutate the state and pending status.
+                if (!nonBlocking)
+                {
+                    try (AutoLock l = _lock.lock())
+                    {
+                        if (_pending || _tryExecutor.tryExecute(this))
+                        {
+                            // This producing thread may block and there is a pending producer available
+                            // so we use EPC: the producer will directly consume the task and then
+                            // race with the pending producer to take over production.
+                            _pending = true;
+                            _state = State.IDLE;
+                            mode = Mode.EXECUTE_PRODUCE_CONSUME;
+                            break;
+                        }
+                    }
+                }
+
+                // To arrive here, either this thread is not allowed to not block
+                // or there were no pending producers available.
+
+                // if the task can be run non blocking
+                if (taskType == Invocable.InvocationType.EITHER)
+                {
+                    // Use PIC to directly consume the task and then resume production.
+                    mode = Mode.PRODUCE_INVOKE_CONSUME;
+                    break;
+                }
+
+                // Otherwise we use PEC: this thread continues to produce and the task is
+                // consumed by the executor
+                mode = Mode.PRODUCE_EXECUTE_CONSUME;
+                break;
+        }
```

Comment:
This is an alternative with some code duplication, but one less test in one case:
```suggestion
        Mode mode;
        // Switch on the invocation type of the produced task
        Invocable.InvocationType taskType = Invocable.getInvocationType(task);
        switch (taskType)
        {
            case NON_BLOCKING:
                // The produced task will never block, so we directly consume the task with PC
                // and then resume production.
                mode = Mode.PRODUCE_CONSUME;
                break;

            case BLOCKING:
                // if this thread may block then we need to take the lock so we can
                // atomically check/mutate the state and pending status.
                if (!nonBlocking)
                {
                    try (AutoLock l = _lock.lock())
                    {
                        if (_pending || _tryExecutor.tryExecute(this))
                        {
                            // This producing thread may block and there is a pending producer available
                            // so we use EPC: the producer will directly consume the task and then
                            // race with the pending producer to take over production.
                            _pending = true;
                            _state = State.IDLE;
                            mode = Mode.EXECUTE_PRODUCE_CONSUME;
                            break;
                        }
                    }
                }

                // Otherwise we use PEC: this thread continues to produce and the task is
                // consumed by the executor
                mode = Mode.PRODUCE_EXECUTE_CONSUME;
                break;

            case EITHER:
                // if this thread may block then we need to take the lock so we can
                // atomically check/mutate the state and pending status.
                if (!nonBlocking)
                {
                    try (AutoLock l = _lock.lock())
                    {
                        if (_pending || _tryExecutor.tryExecute(this))
                        {
                            // This producing thread may block and there is a pending producer available
                            // so we use EPC: the producer will directly consume the task and then
                            // race with the pending producer to take over production.
                            _pending = true;
                            _state = State.IDLE;
                            mode = Mode.EXECUTE_PRODUCE_CONSUME;
                            break;
                        }
                    }
                }

                // To arrive here, either this thread is not allowed to not block
                // or there were no pending producers available.

                // otherwise Use PIC to directly consume the task and then resume production.
                mode = Mode.PRODUCE_INVOKE_CONSUME;
                break;

            default:
                throw new IllegalStateException();
        }
```

Reply:
I've moved on past this alternative.